sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala

```diff
@@ -49,37 +73,120 @@ class PythonMicroBatchStream(
   // from python to JVM.
   private var cachedInputPartition: Option[(String, String, PythonStreamingInputPartition)] = None
 
+  // Store the latest available offset for reporting
+  private var latestAvailableOffset: Option[PythonStreamingSourceOffset] = None
+
   private val runner: PythonStreamingSourceRunner =
     new PythonStreamingSourceRunner(createDataSourceFunc, outputSchema)
   runner.init()
 
   override def initialOffset(): Offset = PythonStreamingSourceOffset(runner.initialOffset())
 
-  override def latestOffset(): Offset = PythonStreamingSourceOffset(runner.latestOffset())
+  /**
+   * Returns the default read limit based on configured options. Supports: maxRecordsPerBatch,
+   * maxFilesPerBatch, maxBytesPerBatch. Falls back to allAvailable if no valid limit is
+   * configured.
+   *
+   * @since 4.2.0
+   */
+  override def getDefaultReadLimit: ReadLimit = {
+    import scala.util.Try
+
+    def parseLong(key: String): Long = {
+      Try(options.get(key).toLong).toOption.filter(_ > 0).getOrElse {
+        throw new IllegalArgumentException(
+          s"Invalid value '${options.get(key)}' for option '$key', must be a positive integer")
+      }
+    }
+
+    def parseInt(key: String): Int = {
+      Try(options.get(key).toInt).toOption.filter(_ > 0).getOrElse {
+        throw new IllegalArgumentException(
+          s"Invalid value '${options.get(key)}' for option '$key', must be a positive integer")
+      }
+    }
+
+    if (options.containsKey("maxRecordsPerBatch")) {
+      val records = parseLong("maxRecordsPerBatch")
+      logInfo(s"Admission control: maxRecordsPerBatch = $records")
+      ReadLimit.maxRows(records)
+    } else if (options.containsKey("maxFilesPerBatch")) {
+      val files = parseInt("maxFilesPerBatch")
+      logInfo(s"Admission control: maxFilesPerBatch = $files")
+      ReadLimit.maxFiles(files)
+    } else if (options.containsKey("maxBytesPerBatch")) {
+      val bytes = parseLong("maxBytesPerBatch")
+      logInfo(s"Admission control: maxBytesPerBatch = $bytes")
+      ReadLimit.maxBytes(bytes)
+    } else {
+      logDebug("No admission control limit configured, using allAvailable")
+      ReadLimit.allAvailable()
+    }
+  }
+
+  override def latestOffset(): Offset = {
+    // Bridge to new signature with default read limit for backward compatibility
+    // Pass null as start offset to maintain backward compatibility with old behavior
+    latestOffset(null, getDefaultReadLimit)
+  }
+
+  /**
+   * Returns the latest offset with admission control limit applied. Also updates the true latest
+   * offset for reporting purposes.
+   *
+   * @param startOffset
+   *   the starting offset, may be null
```

Comment:
nit: let's put in the same line if it doesn't exceed 100 chars

Reply:
nit: Btw, let's not add documentation unless there is noticeable difference from the method doc of base method.